book/src/rfcs/ffadd.md

```diff
@@ -0,0 +1,191 @@
+u_2 &= r_2 + g_2 + k_1\\
+\end{align}
+
+Finally, we perform a range check on the sum $u$, and we would know that $r < f$. 
+
+But now we also have to check that $0\leq r$ or, equivalently, $g\leq u$. Isn't the first check trivial because $r$ is a field element?
+
+
+### Subtractions
+
+Mathematically speaking, a subtraction within a field is no more than an addition over that field. Negative elements are not different from "positive" elements in finite fields (or in any modular arithmetic). To give a general example, the element $-e$ within a field $\mathbb{F}_m$ of order $m$ and $e<m$ is nothing but $m - e$. Nonetheless, for arbitrarily sized elements (not just those smaller than the modulus), the actual field element could be any $c \cdot m - e$, for any multiple $c \cdot m$ of the modulus. Thus, representing negative elements directly as "absolute" field elements may incur in additional computations involving multiplications and thus would result in a less efficient mechanism. 
+
+Alternatively, one can store negative values $-e$ as a negative sign followed by the absolute value $e$. Doing so will affect the way we address foreign field additions in the case that we had $a - b$ or $b - a$ (note that $- a - b$ is just the negation of $a + b$, and nothing changes at the gate level). Specifically, the carries between limbs may not just be bits ($0$ or $1$), but they could also be $-1$. This change only affects the values of $c_0, c_1, c_2$. 
+
+For the second part of the gate, meaning the sum for the range check, the carry bits will remain unchanged. The reason behind this is the fact that $g > 0$ because $2^{264} > f$, and $r$ will be "positive" as well. That means, this sum follows the usual positive-addition carry structure. 
+
+## Gadget
+
+The foreign field gadget will be composed by 4 sets of `RangeCheck` gadgets for witnesses $a, b, r, s$ accounting for 16 rows in total; followed by one row with `ForeignFieldAdd` gate type; and a final `Zero` row. A total of 18 rows with 15 columns in Kimchi.
+
+| Row(s) | Gate type(s) | Witness |
+-|-|-
+0-3 | `multi-range-check` | $a$ |
+4-7 | `multi-range-check` | $b$ |
+8-11 | `multi-range-check` | $r$ |
+12-15 | `multi-range-check` | $s$ |
+16 | `ForeignFieldAdd` |  |
+17 | `Zero` |  |
+
+
+### Layout
+
+For this gate, we need to perform 4 range checks to assert that the limbs of $a, b, r, s$ have a correct size, meaning they fit in $2^{88}$ (and thus, range-checking $a, b, r, s$ for $2^{264}$). Because each of these elements is split into 3 limbs, we will have to use 3 copy constraints between the `RangeCheck` gates and the `ForeignFieldAdd` rows (per element). That amounts to 12 copy constraints. Recalling that Kimchi only allows for the first 7 columns of each row to host a copy constraint, we necessarily have to use 2 rows for the actual addition gate. The layout of these two rows is the following:
+
+|                 | Curr              | Next |
+|-|-|-|
```

Comment:
(btw I use the "markdown all in one" extension on vscode and it will autoformat tables for you)

Reply:
Thanks for this tip!